src/cryptsetup/cryptsetup-token-systemd-tpm2/cryptsetup-token-systemd-tpm2.c

```diff
@@ -0,0 +1,223 @@
+                        policy_hash,
+                        policy_hash_size,
+                        &decrypted_key,
+                        &decrypted_key_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* Before using this key as passphrase we base64 encode it, for compat with homed */
+        r = base64mem(decrypted_key, decrypted_key_size, &base64_encoded);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* free'd automaticaly by libcryptsetup */
+        *password_len = strlen(base64_encoded);
+        *password = TAKE_PTR(base64_encoded);
+
+        return 0;
+}
+
+/*
+ * libcryptsetup callback for memory deallocation of 'password' parameter passed in
+ * any crypt_token_open_* plugin function
+ */
+_public_ void cryptsetup_token_buffer_free(void *buffer, size_t buffer_len) {
+        erase_and_free(buffer);
+}
+
+/*
+ * prints systemd-tpm2 token content in crypt_dump().
+ * 'type' and 'keyslots' fields are printed by libcryptsetup
+ */
+_public_ void cryptsetup_token_dump(
+                struct crypt_device *cd /* is always LUKS2 context */,
+                const char *json /* validated 'systemd-tpm2' token if cryptsetup_token_validate is defined */) {
+
+        bool first;
+        uint32_t i, pcr, pcr_mask;
+        _cleanup_free_ char* base64_blob = NULL, *hex_policy_hash = NULL;
+
+        if (parse_luks2_tpm2_data(json, UINT32_MAX, &pcr_mask, &base64_blob, &hex_policy_hash))
+                return;
+
+        crypt_log_std(cd, "\ttpm2-pcrs:  ");
+        for (i = 0, first = true; i < TPM2_PCRS_MAX; i++) {
+                pcr = pcr_mask & (UINT32_C(1) << i);
+                if (pcr) {
+                        crypt_log_std(cd, "%s%" PRIu32, first ? "" : ", ", pcr);
+                        first = false;
+                }
+        }
+
+        /* well, it's stored in plaintext json metadata anyway... */
+        crypt_log_std(cd, "\n\ttmp2-blob:  %s\n", base64_blob);
+
+        crypt_log_std(cd, "\ttmp2-policy-hash: %s\n", hex_policy_hash);
+}
+
+/*
+ * Note:
+ *   If plugin is available in library path, it's called in before following libcryptsetup calls:
+ *
+ *   crypt_token_json_set, crypt_dump, any crypt_activate_by_token_* flavour
+ */
+_public_ int cryptsetup_token_validate(
+                struct crypt_device *cd, /* is always LUKS2 context */
+                const char *json /* contains valid 'type' and 'keyslots' fields. 'type' is 'systemd-tpm2' */) {
+
+        int r;
+        JsonVariant *w, *e;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ void *blob = NULL, *policy_hash = NULL;
+        size_t blob_size = 0, policy_hash_size = 0;
+
+        r = json_parse(json, 0, &v, NULL, NULL);
+        if (r < 0) {
+                crypt_log_dbg(cd, "Could not parse " TOKEN_NAME " json object%s.",
+                              r == -ENOMEM ? " (not enough memory)" : "");
```

Comment:
maybe add:

```c
#define crypt_log_debug_errno(cd, e, ...) \
        ({  \
                int _e = abs(e), _s = errno; \
                errno = _e; \
                crypt_logf(cd, CRYPT_LOG_DEBUG, __VA_ARGS__) \
                errno = _s; \
                -_e; \
})
```

Such a macro would make logging a lot more like our usually logging: i.e. %m will resolve to strerror(errno) in a thread-safe way. You could then rewrite this call here to:

```c
crypt_log_debug_errno(cd, r, "Could not parse "TOKEN_NAME " JSON object: %m");
```

Which is a log prettier I think and works reasonably for all kinds of errors